### PR TITLE
PP-10964 Handle Handle Stripe card_decline_rate_limit_exceeded error code as authorisation rejected

### PR DESCRIPTION
**Pull request by AbdirizakIdris**
## WHAT YOU DID
- Updated the Stripe authorisation error handling logic to handle errors with type ``` invalid_request_error ``` and code ```
card_decline_rate_limit_exceeded ``` by updating the state to ```AUTHORISATION_REJECTED ```

### CODE REVIEW

**Review by SandorArpa — DISMISSED** [review]
LGTM

File: src/main/java/uk/gov/pay/connector/gateway/stripe/json/StripeAuthorisationFailedResponse.java [hunk 1]
```diff
@@ -44,7 +44,9 @@ public AuthoriseStatus authoriseStatus() {
            (https://stripe.com/docs/api/errors#errors-card_error)
          */
         if (errorResponse != null && errorResponse.getError() != null
-                && "card_error".equals(errorResponse.getError().getType())) {
+                && ("card_error".equals(errorResponse.getError().getType())
+                || "invalid_request_error".equals(errorResponse.getError().getType())
```
Comment (kbottla) [inline src/main/java/uk/gov/pay/connector/gateway/stripe/json/StripeAuthorisationFailedResponse.java]: The condition should be as below:

```
if (errorResponse != null && errorResponse.getError() != null
                && ("card_error".equals(errorResponse.getError().getType())
                || ("invalid_request_error".equals(errorResponse.getError().getType()) 
                && "card_decline_rate_limit_exceeded".equals(errorResponse.getError().getCode())))) {
```

For error type `invalid_request_error`, we can get different error codes. Only for `card_decline_rate_limit_exceeded` error code, payment status should be rejected.

**Review by kbottla — CHANGES_REQUESTED** [review]

File: src/test/java/uk/gov/pay/connector/gateway/stripe/StripeCaptureHandlerTest.java [hunk 2]
```diff
@@ -215,9 +215,12 @@ void shouldCaptureWithoutFee_ifCollectFeeSetToFalse() throws Exception {
 
     @Test
     void shouldNotCaptureIfPaymentProviderReturns4xxHttpStatusCode() throws Exception {
-        GatewayErrorException exception = new GatewayErrorException("Unexpected HTTP status code 402 from gateway", load(STRIPE_ERROR_RESPONSE), SC_UNAUTHORIZED);
+        String errorResponse = load(STRIPE_ERROR_RESPONSE);
```
Comment (kbottla) [inline src/test/java/uk/gov/pay/connector/gateway/stripe/StripeCaptureHandlerTest.java]: Can simplify this
```suggestion
        String errorMessage = load(STRIPE_ERROR_RESPONSE)
                              .replace("{{code}}", "resource_missing");
```

File: src/test/java/uk/gov/pay/connector/gateway/stripe/StripeCaptureHandlerTest.java [hunk 3]
```diff
@@ -243,7 +246,9 @@ void shouldNotCaptureIfPaymentProviderReturns4XXOnTransfer() throws Exception {
         when(gatewayCaptureResponse.getEntity()).thenReturn(load(STRIPE_PAYMENT_INTENT_CAPTURE_SUCCESS_RESPONSE));
         when(gatewayClient.postRequestFor(any(StripeCaptureRequest.class))).thenReturn(gatewayCaptureResponse);
 
-        GatewayErrorException exception = new GatewayErrorException("Unexpected HTTP status code 402 from gateway", load(STRIPE_ERROR_RESPONSE), SC_UNAUTHORIZED);
+        String errorResponse = load(STRIPE_ERROR_RESPONSE);
```
Comment (kbottla) [inline src/test/java/uk/gov/pay/connector/gateway/stripe/StripeCaptureHandlerTest.java]: Same as above


```suggestion
        String errorMessage = load(STRIPE_ERROR_RESPONSE)
                              .replace("{{code}}", "resource_missing");
```

File: src/test/java/uk/gov/pay/connector/gateway/stripe/StripeRefundHandlerTest.java [hunk 4]
```diff
@@ -179,7 +179,9 @@ private void mockRefundSuccess() throws GatewayException.GenericGatewayException
     }
 
     private void mockRefund4xxResponse() throws GatewayException.GenericGatewayException, GatewayErrorException, GatewayException.GatewayConnectionTimeoutException {
-        GatewayErrorException gatewayClientException = new GatewayErrorException("Unexpected HTTP status code 402 from gateway", load(STRIPE_ERROR_RESPONSE), 402);
+        String errorResponse = load(STRIPE_ERROR_RESPONSE);
```
Comment (kbottla) [inline src/test/java/uk/gov/pay/connector/gateway/stripe/StripeRefundHandlerTest.java]: Same as above for both test changes here

**Review by kbottla — DISMISSED** [review]

**Review by kbottla — APPROVED** [review]

**Pull request merged**